addons/mail/models/mail_mail.py

```diff
@@ -214,22 +214,33 @@ def _send_prepare_values(self, partner=None):
         }
         return res
 
-    def _split_by_server(self):
-        """Returns an iterator of pairs `(mail_server_id, record_ids)` for current recordset.
+    def _split_by_mail_configuration(self):
+        """Group the <mail.mail> based on their "email_from" and their "mail_server_id".
 
-        The same `mail_server_id` may repeat in order to limit batch size according to
-        the `mail.session.batch.size` system parameter.
+        The <mail.mail> will have the "same sending configuration" if they have the same
+        mail server or the same mail from. For performance purpose, we can use an SMTP
+        session in batch and therefore we need to group them by the parameter that will
+        influence the mail server used.
+
+        The same "sending configuration" may repeat in order to limit batch size
+        according to the `mail.session.batch.size` system parameter.
+
+        Return iterators over
+            mail_server_id, email_from, Records<mail.mail>.ids
         """
+        mail_values = self.read(['id', 'email_from', 'mail_server_id'])
+
         groups = defaultdict(list)
-        # Turn prefetch OFF to avoid MemoryError on very large mail queues, we only care
-        # about the mail server ids in this case.
-        for mail in self.with_context(prefetch_fields=False):
-            groups[mail.mail_server_id.id].append(mail.id)
+        for values in mail_values:
```

Comment:
Here, basically we are going to make groups based on (mail_server_id, from). However we have a lot of use cases where mail_server_id will not be given but we could have a lot of different email_from (e.g. doing a mass_mailing with email_from being jinja-based and therefore could be based on a document's responsible).

When having a mail_server_id set, in ``connect`` you seem to discard smtp_from as we somehow "force" the server to use.

Couldn't we call use the ``_find_mail_server`` heuristic to save time here and prepare groups directly with coherent values ? Global idea would be
```
1/ group by mail_server_id (if forced server, we don't need to classify based on from, unless I am mistaken)
2/ for all other mails:
  * find its preferred server
  * add mails to already-existing groups or create new groups, based on mailserver_id
3/ send, now with always a server set
```

Indeed I don't think it is ``connect`` job to do it, and it will be inefficient (as it could be email-per-email). I think preparing groups then calling connect would be more performant. And it makes sense to delegate this job to a global "send" method on mail.mail.

What do you think ?

Reply:
@tde-banana-odoo 
> add mails to already-existing groups or create new groups, based on mailserver_id

This is not really possible, because even if they use the same mail server, they might use a different smtp_from for the authentication, (for the case where we have different email_from which will in the end use the same outgoing server))

So we will always need to create new group, and it won't change that much if we call `find_mail_server` before except that it will complexify this part of the code

Reply:
@tde-banana-odoo It's a bit tricky, because even if some emails used the same mail server, they might use a different smtp_from for the authentication (connect method of `ir.mail_server`)...

What I can do to improve this in most cases is to encapsulate the email later, so in practice, if you have a mail server configured for the notification email, all the emails will be in the same group. (and the FROM for the SMTP header will be encapsulated later).

So now
- I do not encapsulate the email in "find_mail_server"
- I encapsulate them in `_prepare_email_message`

What do you think ?

Reply:
(I pushed the change so you can see)

Reply:
@tde-banana-odoo The tricky thing is that in some cases, some emails could use the same mail server, but with a different login (smtp_from).

But I optimized it for the most common case, when we use a notification email. I encapsulate the email later (just before the sending), so I can group the <mail.mail> based on the mail server and the smtp from (login to the mail server).

I updated the test `test_mail_mail_sending_batch` so you can see how the <mail.mail> will be grouped.